01_fib/undrewb/main_test.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := strconv.Quote("0\n1\n1\n2\n3\n5\n8\n13\n")
+	actual := strconv.Quote(buf.String())
+
+	if expected != actual {
+		t.Errorf("Unexpected result in main()")
+	}
+}
+
+func TestFibNegativeOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	fib(-7)
+
+	expected := strconv.Quote("0\n1\n-1\n2\n-3\n5\n-8\n13\n")
+	actual := strconv.Quote(buf.String())
+
+	if expected != actual {
+		t.Errorf("Unexpected result in negative fib")
+	}
+}
+
+var fibTests = []struct {
+	n        int // input
+	expected int // expected result
+}{
+	{1, 1},
+	{2, 1},
+	{3, 2},
+	{4, 3},
+	{5, 5},
+	{6, 8},
+	{7, 13},
+}
+
+func TestFibonacci(t *testing.T) {
+	for _, tt := range fibTests {
+		actual := fibonacci(tt.n)
+		if actual != tt.expected {
+			t.Errorf("fibonacci(%d): expected %d, actual %d", tt.n, tt.expected, actual)
+		}
+	}
+}
+
+var negafibTests = []struct {
+	n        int // input
+	expected int // expected result
+}{
+	{1, 1},
+	{2, -1},
+	{3, 2},
+	{4, -3},
+	{5, 5},
+	{6, -8},
+	{7, 13},
+}
+
+func TestNegaFibonacci(t *testing.T) {
+	for _, tt := range negafibTests {
+		actual := negafibonacci(tt.n)
+		if actual != tt.expected {
+			t.Errorf("negafibonacci(%d): expected %d, actual %d", tt.n, tt.expected, actual)
+		}
+
+	}
+}
+
+func TestFibonacciZero(t *testing.T) {
+	value := fibonacci(0)
+	if value != 0 {
+		t.Errorf("fibonacci zero test failed\n")
+	}
+}
+func TestFibonacciOverflow(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	fib(21)
+
+	expected := strconv.Quote("I dont go past twenty\n")
+	actual := strconv.Quote(buf.String())
+
+	if expected != actual {
+		t.Errorf("Unexpected handling of overflow in Fib()")
+	}
+}
+
+func TestSignPositive(t *testing.T) {
+	if sign(2) != 1 {
+		t.Errorf("test sign positive failed\n")
+	}
+}
+
+func TestSignNegativeSix(t *testing.T) {
```

Comment:
nit: I would fold these into a single test with a loop covering a range of negative even numbers and another covering odd numbers. Make sure `t.Errorf` reports the `n` that caused the failure and the value produced by `sign`.

Reply:
In progress.

Reply:
Done